### PR TITLE
repair: drop timeout from table_sync_and_check

### PR DESCRIPTION
**Pull request by Deexie**
Delete 10s timeout from read barrier in table_sync_and_check, so that the function always considers all previous group0 changes.

Fixes: #18490.

- [x] ** Backport reason (please explain below if this patch should be backported or not) **
Should be backported to 5.2 and 5.4 as they use timeout

### CODE REVIEW

**Review by denesb — APPROVED** [review]

**Comment by denesb** [comment]
@kbr-scylla please review.

**Comment by kbr-scylla** [comment]
I would like the tests from https://github.com/scylladb/scylladb/issues/18490 to be reran to confirm that this PR indeed fixes the issue -- **before** we merge this.

Alternatively, it needs a reproducer in test.py or dtest.

Merging without proper testing is why issues like https://github.com/scylladb/scylladb/issues/18490 and https://github.com/scylladb/scylladb/issues/18452 are happening after https://github.com/scylladb/scylladb/pull/17231 was merged.

I don't want to see the same situation as before, where we replace one error with another.


Also @tgrabiec 

> What's the reasoning behind 10s timeout? I think repair doesn't have to time out on group0 barrier at all, provided that it reliably eventually completes. If user wants to give up, he should cancel repair.

I have problems with this approach:
- is repair actually cancellable? Do we have an API? (Other than shutting down nodes -- which is very bad UX for cancelling things)
- how does the user know that repair hanged? The timeout **is** a way to inform the user that something is wrong and repair cannot continue. If we hang forever, there will be no way for the user to know anything, unless we introduce some reporting mechanism (for example, a background task that would log a message for the user to inform -- we hanged waiting for schema sync for more than X etc., consider canceling)


The second point alone is enough for me to reject this patch.

I have two alternatives to propose:
- keep using a timeout, simply increase it.
- If you don't want to use a timeout, there should be a mechanism to inform the user that there's a problem, we cannot silently hang forever.

**Comment by kbr-scylla** [comment]
Also, will this fix https://github.com/scylladb/scylladb/issues/18452? Looks like it could -- again **testing is needed**

**Comment by kbr-scylla** [comment]
> If we hang forever, there will be no way for the user to know anything, unless we introduce some reporting mechanism (for example, a background task that would log a message for the user to inform -- we hanged waiting for schema sync for more than X etc., consider canceling)

You can find an example of this approach in service/raft/raft_group0.cc, in the upgrade-cluster-to-use-Raft procedure implementation. This procedure:
- requires full cluster availability, so it's possible for the procedure to fail with just a single down node
- there is no sensible way to cancel it without requiring some kind of manual user intervention -- so the only sensible automatic way for dealing with problems is to keep retrying forever (until full cluster availability is restored, or manual recovery procedure is performed)

Because of these properties, I implemented a background task that waits for the procedure to complete, but if it doesn't after 1 minute, it starts reporting a message to the log (every minute) that the procedure has hanged and manual intervention may be required.

```cpp
static auto warn_if_upgrade_takes_too_long() {
    auto as = std::make_unique<abort_source>();
    auto task = [] (abort_source& as) -> future<> {
        static constexpr auto warn_period = std::chrono::minutes{1};

        while (!as.abort_requested()) {
            try {
                co_await sleep_abortable(warn_period, as);
            } catch (const sleep_aborted&) {
                co_return;
            }

            upgrade_log.warn(
                "Raft upgrade procedure taking longer than expected. Please check if all nodes are live and the network is healthy."
                " If the upgrade procedure does not progress even though the cluster is healthy, try performing a rolling restart of the cluster."
                " If that doesn't help or some nodes are dead and irrecoverable, manual recovery may be required."
                " Consult the relevant documentation: {}", raft_upgrade_doc);
        }
    }(*as);

    return defer([task = std::move(task), as = std::move(as)] () mutable {
        // Stop in background.
        as->request_abort();
        (void)std::move(task).then([as = std::move(as)] {});
    });
}
```

**Comment by denesb** [comment]
> I have two alternatives to propose:
> 
>     * keep using a timeout, simply increase it.

@tgrabiec seems to favor keeping the timeout: https://github.com/scylladb/scylladb/pull/18613#issuecomment-2123435326

**Comment by denesb** [comment]
@Deexie let's increase the timeout instead of eliminating it, lets increase it to say 5 minutes. That should be enough for a schema change to go through and it will not cause repair to "hang forever".

@kbr-scylla as for testing: @Deexie tried to run the reproducer and after 2 weeks of trying, with a lot of help from QA infra team, we got 0 runs, not even the AMI was built successfully, for various permission and other problems.
If we merged it 3 weeks ago, it would have got a lot of longevity testing in the meanwhile. I agree that we should test as much as we can and as early as we can, but we have to be pragmatic, especially with such hard to reproduce problems as this, where you need these specific circumstances to reproduce:
* Index/table dropped while being repaired or RBNO going on.
* Table drop triggers a schema change which is expensive enough to hit the 10second group0 barrier timeout.

According to QA this reproduced quite rarely on SCT to begin with.

**Comment by kbr-scylla** [comment]
> @kbr-scylla as for testing: @Deexie tried to run the reproducer and after 2 weeks of trying, with a lot of help from QA infra team, we got 0 runs, not even the AMI was built successfully, for various permission and other problems.

Damn. I don't remember this to be so bad, when I was working on CDC 4 years ago I was constantly running longevities to test CDC-based replication, there was some trouble with initial setup and a bunch of issues that I reported on SCT repo but eventually it worked. Regarding building Scylla artifacts, I did build AMIs at some point, but for that particular CDC case, I think I was not building AMIs but just the *rpm/*deb packages, which should be easy to do today with the byo_build_tests_dtest

@Deexie Please escalate wherever necessary. We cannot leave the current state as it is.
cc @scylladb/qa-maintainers @scylladb/releng 

> If we merged it 3 weeks ago, it would have got a lot of longevity testing in the meanwhile.

But that is not the solution to the problems you encountered, that would be avoidance. We must fix SCT and our Jenkins jobs and anything else that is required so that **any** developer can run these tests whenever they need to.

If our tools are broken we need to fix them
This will be a problem for any developer who finds the need to run longevity to test their stuff :(

---

Other than that, even if we consider the lack of testing to be not a blocker, I still had my complaints regarding the approach itself which should be addressed before merging.

**Comment by mykaul** [comment]
@denesb, @Deexie - please let me know if you get stuck on such issues in the future. We need to get better at this.
@roydahan - let's see what could we have done differently here.

**Comment by denesb** [comment]
> But that is not the solution to the problems you encountered, that would be avoidance. We must fix SCT and our Jenkins jobs and anything else that is required so that **any** developer can run these tests whenever they need to and it should be **easy**
> 
> If our tools are broken we need to fix them This will be a problem for any developer who finds the need to run longevity to test their stuff :(

In theory I agree, but I don't think this is working. If a tool takes days of communication back-and-forth and hours on end of effort to setup and finally run, anybody exposed to this, will avoid it like the plague in the future, as well as anybody who hears of this experience.
If we really want developers to use these tools, the developers of said tool should do some regular and minimal testing of the on-boarding procedure as well as of the use-case a regular core developer is expected to use -- bring a branch of their own. Currently, the tool works fine for the use-cases QA uses it for, for people who have done all their setup a long time ago. But for developers it is a very rough experience.
This is a similar problem to that we had in scylla-core, regarding setting up a build environment. We solved it with the frozen toolchain, and now anybody can get from `git clone` to compiling in just a few minutes.

**Comment by fruch** [comment]
> > But that is not the solution to the problems you encountered, that would be avoidance. We must fix SCT and our Jenkins jobs and anything else that is required so that **any** developer can run these tests whenever they need to and it should be **easy**
> > If our tools are broken we need to fix them This will be a problem for any developer who finds the need to run longevity to test their stuff :(
> 
> In theory I agree, but I don't think this is working. If a tool takes days of communication back-and-forth and hours on end of effort to setup and finally run, anybody exposed to this, will avoid it like the plague in the future, as well as anybody who hears of this experience. If we really want developers to use these tools, the developers of said tool should do some regular and minimal testing of the on-boarding procedure as well as of the use-case a regular core developer is expected to use -- bring a branch of their own. Currently, the tool works fine for the use-cases QA uses it for, for people who have done all their setup a long time ago. But for developers it is a very rough experience. This is a similar problem to that we had in scylla-core, regarding setting up a build environment. We solved it with the frozen toolchain, and now anybody can get from `git clone` to compiling in just a few minutes.

issue here is more then a frozen toolchain (which SCT has), 
we are testing again and again the on-boarding process, and any new member of joined in to the testing team, go it it up and running within the same day, as part of their onboarding process.

at the end is misalignment of expectations, 

SCT on it's own can't solve permission issues to AWS or Jenkins, but one will still bash the tool for those.
SCT on it's own can't build AMIs, or any of scylla core artifacts, but still one would bash the tools, for no being able todo that.

people should communicate their needs and the issues they are running into
if there don't we don't know about it and can prioritize any fixes or improvements.

**Comment by kbr-scylla** [comment]
The use case for Core devs is "I have a branch, I want to run longevity against it"
IIRC it was also something that was selected often in the survey you did

the problem with this use case is that:
- it happens rarely
- once it happens, the dev doesn't know (even if they did, they don't remember) how to get from the branch to something that you can run SCT against (i.e. how to build all the necessary artifacts)
- once they finally have the artifacts (whether it's an AMI, or *rpm/*deb packages, or whatever), they actually have to run SCT

We should have Jenkins jobs for each of these steps that are **easy** to use, i.e., you put link to your fork on github and the branch into the "build artifact" job parameters, click run, and you get artifact link. Then you put your artifact link into the "run SCT" job parameters, click run, and you get reports with logs, dashboards etc.

Even better if there was a combined job which does both of these steps.

**Comment by kbr-scylla** [comment]
> The use case for Core devs is "I have a branch, I want to run longevity against it"

by branch I meant a branch of a fork of `scylladb/scylladb` or `scylladb/scylla-enterprise`

even better if it's also possible to provide fork+branch for `scylladb/scylla-cluster-tests` too.

**Comment by kbr-scylla** [comment]
And just as I finished writing those comments, I saw a status email from Testing Infra team, that @vponomaryov is working on such a job. Great!

**Comment by denesb** [comment]
> The use case for Core devs is "I have a branch, I want to run longevity against it"

A much more generic and useful way is to say: here is a package that I built, I want to run longevity against it. This is even better then "I have a branch", because I have control over the build mode, build options and even the installation if I want to be fancy.
I don't care which package this is, as long as I can just type `ninja dist-something` and I get the package as a result.

**Comment by fruch** [comment]
> > The use case for Core devs is "I have a branch, I want to run longevity against it"
> 
> A much more generic and useful way is to say: here is a package that I built, I want to run longevity against it. This is even better then "I have a branch", because I have control over the build mode, build options and even the installation if I want to be fancy. I don't care which package this is, as long as I can just type `ninja dist-something` and I get the package as a result.

at time, the AMIs could been built like that, no one did built them, only specific individuals
since then we have the scylla-machine-image repo, that is handling all of it.

let give you a different example, do you know how to build the scylla docker image  ? it's not depended on any external repo,
but still I don't think there's a ninja command for it... (or at least wasn't last time I was looking)

**Comment by tgrabiec** [comment]
> > What's the reasoning behind 10s timeout? I think repair doesn't have to time out on group0 barrier at all, provided that it reliably eventually completes. If user wants to give up, he should cancel repair.
> 
> I have problems with this approach:
> 
> * is repair actually cancellable? Do we have an API? (Other than shutting down nodes -- which is very bad UX for cancelling things)

Yes, there's a RESTful API for that.

> * how does the user know that repair hanged? **The timeout is a way to inform the user that something is wrong** and repair cannot continue. If we hang forever, there will be no way for the user to know anything, unless we introduce some reporting mechanism (for example, a background task that would log a message for the user to inform -- we hanged waiting for schema sync for more than X etc., consider canceling)

There are many places where repair can theoretically hang, not just this one. I don't like timeouts for detecting this because, as we have seen here, it is prone to false positives. In the best case it's frustrating to the user, but there's also no guarantee that it won't timeout again when you retry it (liveness issues). 

In the absence of bugs, if the barrier hangs now, the barrier on retry will also hang. So retrying doesn't solve anything. The problem is not with repair but with group0, so I think group0 should report the stuckness and admin should fix it, and then anything which is blocked on group0 will be unblocked. In this case it's better to fix it centrally for all users of the barrier.

**Comment by kbr-scylla** [comment]
Well, any kind of reporting is prone to false positives. We might report stuckness when it will get automatically unstuck in 1 second from now.

> The problem is not with repair but with group0, so I think group0 should report the stuckness and admin should fix it, and then anything which is blocked on group0 will be unblocked. In this case it's better to fix it centrally for all users of the barrier.

You just reminded me that indeed, group 0 did provide a generic solution, and that solution uses timeouts. https://github.com/scylladb/scylladb/pull/17590

Group 0 took the timeout solution because group 0 operations should never be so expensive to take significant amount of time. If a group 0 operation takes longer than, say, a minute, this is indication of a deeper problem, and solving it by waiting more is just hiding the real problem under the rug. The real problem could be something like:
- we're waiting on the wrong semaphore that should be dedicated to user requests only (https://github.com/scylladb/scylladb/issues/18863)
- or getting stuck at memtable flushes for memtables storing user data because we use the wrong memory pool (https://github.com/scylladb/scylladb/issues/15622)
 
etc.

IMO better to detect those problems as early as possible and fix them, rather than using 1 hour timeouts or no timeouts at all and then wondering why certain internal operations which should be instantaneous are taking forever.

---

Anyway I just reviewed the code deeper and:
- `mm.get_group0_barrier().trigger(mm.get_abort_source());` uses `migration_manager::_group0_barrier`
- `_group0_barrier` is constructed like this:
```cpp
        , _group0_barrier(this_shard_id() == 0 ?
            std::function<future<>()>([this] () -> future<> {
                // This will run raft barrier and will sync schema with the leader
                (void)co_await start_group0_operation();
            }) :
            std::function<future<>()>([this] () -> future<> {
                co_await container().invoke_on(0, [] (migration_manager& mm) -> future<> {
                    // batch group0 raft barriers
                    co_await mm._group0_barrier.trigger();
                });
            })
        )
```
- `start_group0_operation` is implemented as this:
```cpp
future<group0_guard> migration_manager::start_group0_operation() {
    assert(this_shard_id() == 0);
    return _group0_client.start_operation(&_as, raft_timeout{});
}
```
- we see that `raft_timeout{}` is passed. `raft_group0_client::start_operation` underneath uses
```cpp
            co_await _raft_gr.group0_with_timeouts().read_barrier(as, timeout);
```

which will use the passed `raft_timeout{}`... which means we use the default timeout for group 0 -- 1 minute.
```cpp
    lowres_clock::duration default_op_timeout = std::chrono::minutes(1);
```
(from `create_server_for_group0`)

So my previous review was wrong -- **the PR as it is now already uses the built-in 1 minute timeout for group 0**.

**Comment by denesb** [comment]
> So my previous review was wrong -- **the PR as it is now already uses the built-in 1 minute timeout for group 0**.

Does that mean that removing the `abort_on_expiry`, as this PR does, still leaves us with a built-in timeout in group0? So this PR is good-to-go (pending the dreaded testing).

**Comment by kbr-scylla** [comment]
Yes

**Comment by tgrabiec** [comment]
> Well, any kind of reporting is prone to false positives. We might report stuckness when it will get automatically unstuck in 1 second from now.

There's reporting and there's failing the operation. False positives in reporting are file, it's just information. Failing the operation is less fine. We're making a benign performance issue into a more serious issue, which causes cascading failure. This makes the system more fragile, we should strive to keep the system anti-fragile.

We've seen several field issues where node operations were timing out because of current conditions (e.g. large token metadata) and we could not make progress because of that. It doesn't mean we shouldn't detect issues, but we shouldn't let those issues have higher impact on the system that they need to have.

We could report stuck slow group0 command application in logs without failing the barrier. It's a common pattern, we already have CPU stall detector, topology barrier stall detector. None of these fail the underlying operation, and you wouldn't want them to.

**Comment by kbr-scylla** [comment]
> We're making a benign performance issue into a more serious issue

It's not benign. There's a place where **data path blocks on group 0 barrier** -- `migration_manager::get_schema_for_write`. If we do an ALTER, then if group 0 blocks for 1 second -- we have a latency spike of 1 second on the data path. Group 0 blocks for 2 seconds -- writes time out. Group 0 blocks for 5 seconds -- reads time out. (Assuming default timeouts.)

So I prefer to catch those group 0 performance issues ASAP.

> We could report stuck slow group0 command application in logs without failing the barrier. It's a common pattern, we already have CPU stall detector, topology barrier stall detector. None of these fail the underlying operation, and you wouldn't want them to.

It's nice that we report, but what then? Who reads these reports? If they do, do they investigate them or at least create issues and plan for eventually fixing?

For example this "topology version held" message. I saw them in some test runs -- and I have no idea what to do with them. Should I report them as bugs? Should I investigate them (but how? is there any documentation written which would guide me?) Are they signifying some serious problems? What do they actually mean?

Specific example: https://github.com/scylladb/scylladb/issues/17938#issuecomment-2010109956
you sent a fix which closed this issue, where I asked about this "topology version held" message: https://github.com/scylladb/scylladb/pull/19137#issuecomment-2152258677, and you answered that it's probably an unrelated issue.

Then the original issue was closed. I didn't see a follow-up opened regarding the topology stall. It's there, but it was ignored. It will be forgotten, because it appeared in an issue that is no closed, and there are not many things less interesting than closed issues.

At least failures force developers to act, investigate, sometimes find serious bugs and fix them.

**Comment by kostja** [comment]
@kbr-scylla catch performance issues != abort writes. I agree with @denesb we need a 5-10m timeout. If you want a warning in the log or a monitoring metric for that, it's another point.

There is no point timing out schema sync at all.

**Comment by nyh** [comment]
> > A much more generic and useful way is to say: here is a package that I built, I want to run longevity against it. This is even better then "I have a branch", because I have control over the build mode, build options and even the installation if I want to be fancy. I don't care which package this is, as long as I can just type `ninja dist-something` and I get the package as a result.
> 
> at time, the AMIs could been built like that, no one did built them, only specific individuals since then we have the scylla-machine-image repo, that is handling all of it.
> 
> let give you a different example, do you know how to build the scylla docker image ? it's not depended on any external repo, but still I don't think there's a ninja command for it... (or at least wasn't last time I was looking)

In my view, the only solution to this problem is **documentation**. In the past, I did build Scylla dockers, relocatable packages _and_ AMIs quite successfully, using various scripts we had for that. But not having done this for many months - I don't remember ANYTHING.  We need an easy do find document in docs/dev on how to build different types of images. I remembered there was such a document about document and after a recursive grep, found it in: dist/docker/debian/README.md (I wrote half of it, and already in d7474ddff3d5b49f03848fccdc78ad758e95ab4c complained about its bad placement :-)). That's very hard to find. It should be something like docs/dev/docker.md or something.

In my opinion, magic Jenkins jobs which do these builds for me given a branch, are **not** what I want as a developer. First, it doesn't give me any control over how the build is done or be able to fix something that breaks. But perhaps more importantly - we would also need documentation about that! @fruch maybe you and two other people in the company know all the nice Jenkins jobs that do all sorts of clever things like builds and test. I don't. I never remember where these jobs are, or if I find one - did I find the right one? What do its parameters mean? Again, we need a document docs/dev/jenkins. I find it hard to accept that we keep saying that Jenkins is the solution to all developers' problems - but Jenkins isn't mentioned even once anywhere in any docs/dev document. How are new developers (or old developers who forgot) supposed to learn which Jenkins jobs we have and how to use them?

**Pull request closed without merging**
